### PR TITLE
feat: file picker added and access token generation

### LINKED ISSUE #20163: [Feature]: GSheets Request Access Token
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

_When frontend engineer makes API request to get the access token, I should see loading icon till the request is complete, so that user cannot perform any action on datasource page until response is received_

1. When frontend engineer makes API request to fetch access token:
    1. user should see loading icon till request is complete
    2. Once request is complete, hide loading icon and If response is,
        1. valid, show file picker popup
        2. invalid, show appropriate error message on datasource page, update OAuth2 status in datasource configuration to unsuccessful.

### Why should this be worked on?

-

### PR DESCRIPTION
**Pull request by sneha122**
## Description

This PR includes following changes:

- In case of limiting google sheet access project, when user selects specific sheets as an option, they should be shown file picker UI once the authorisation is complete, In this file picker UI, users can select the google sheet files that they want to use with appsmith application and allow access to only those files.
- This PR contains the changes for file picker UI and updating datasource auth state based on the files selected by user.

TL;DR
Steps to test this PR:

- Create Google Sheet datasource
- In the datasource config form, select specific sheets as an option from the scope dropdown.
- Click on save and authorise
- This will take you to google oauth process

Screenshot 2023-02-20 at 1 24 24 PM

- Select the google account
- This will take you to google oauth2 consent screen

Screenshot 2023-02-20 at 1 24 55 PM

- Click on allow for all requested permissions
- This will take you back to appsmith's datasource config page in view mode and load the file picker UI
Screenshot 2023-02-20 at 1 25 47 PM
- Select the files that you want to share with appsmith app
- Click on select
- You should see the new query button in enabled state, as datasource authorisation is complete

Screenshot 2023-02-20 at 1 27 28 PM

- In case you select cancel on google oauth2 consent screen, you should error message on datasource config page with new query button being disabled

Screenshot 2023-02-20 at 1 28 49 PM

- In case you do give all the permissions but do not select any files in google file picker, then also you should see error message on datasource config page with new query button disabled.

Fixes #20163, #20290, #20160, #20162

Media
> A video or a GIF is preferred. when using Loom, don’t embed because it looks like it’s a GIF. instead, just link to the video

## Type of change

> Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Manual

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/src/pages/Editor/SaaSEditor/DatasourceForm.tsx [hunk 1]
```diff
@@ -441,7 +444,7 @@ const mapStateToProps = (state: AppState, props: any) => {
     ...pagePermissions,
   ]);
 
-  const featureFlags: FeatureFlags = selectFeatureFlags(state);
+  const gsheetToken = state.entities.datasources.gsheetToken;
```
Comment (Irongade) [inline app/client/src/pages/Editor/SaaSEditor/DatasourceForm.tsx:447]: Can we use a proper selector here?

File: app/client/src/pages/common/datasourceAuth/index.tsx [hunk 2]
```diff
@@ -148,9 +157,18 @@ function DatasourceAuth({
 
   const pageId = (pageIdQuery || pageIdProp) as string;
   const [confirmDelete, setConfirmDelete] = useState(false);
+
+  const [scriptLoadedFlag] = useState<boolean>(
```
Comment (Irongade) [inline app/client/src/pages/common/datasourceAuth/index.tsx:161]: What happens if the googleApisLoaded is false here? that is, if the page is loaded and the google apis are not loaded some how? We can either prompt the user to reload the page or somehow try to reload the script again (this can be done on another PR).

File: app/client/src/pages/common/datasourceAuth/index.tsx [hunk 3]
```diff
@@ -285,6 +303,49 @@ function DatasourceAuth({
     }
   };
 
+  useEffect(() => {
+    // This loads the picker object in gapi script
+    if (!!gsheetToken) {
```
Comment (Irongade) [inline app/client/src/pages/common/datasourceAuth/index.tsx:308]: check if gapi exists before trying to call load function. i.e (!!gsheetToken && !!gapi)

File: app/client/src/pages/common/datasourceAuth/index.tsx [hunk 4]
```diff
@@ -285,6 +303,49 @@ function DatasourceAuth({
...
+        await gapi.client.load(
+          "https://www.googleapis.com/discovery/v1/apis/drive/v3/rest",
+        );
+        setPickerInitiated(true);
+      });
+    }
+  }, [scriptLoadedFlag, gsheetToken]);
+
+  useEffect(() => {
+    if (!!gsheetToken && scriptLoadedFlag && pickerInitiated) {
```
Comment (Irongade) [inline app/client/src/pages/common/datasourceAuth/index.tsx:319]: check that google variable exists before calling createPicker

File: app/client/src/sagas/DatasourcesSagas.ts [hunk 5]
```diff
@@ -1152,6 +1163,30 @@ function* initializeFormWithDefaults(
...
+    yield put({
+      type: ReduxActionTypes.SET_GSHEET_TOKEN,
+      payload: {
+        gsheetToken: "",
+      },
+    });
+
+    if (action === "cancel") {
+      const datasource: Datasource = yield select(getDatasource, datasourceId);
+      _.set(
```
Comment (Irongade) [inline app/client/src/sagas/DatasourcesSagas.ts:1180]: can we only import set function, rather than the entire lodash functions to reduce import size?

File: app/client/src/sagas/DatasourcesSagas.ts [hunk 6]
```diff
@@ -1152,6 +1163,30 @@ function* initializeFormWithDefaults(
...
+        datasource,
+        "datasourceConfiguration.authentication.authenticationStatus",
+        "FAILURE",
```
Comment (Irongade) [inline app/client/src/sagas/DatasourcesSagas.ts:1183]: Can this value be made an enum or can we create a type/interface for this.

File: app/client/src/pages/common/datasourceAuth/index.tsx (hunk 2 above)
Comment (sneha122) [reply app/client/src/pages/common/datasourceAuth/index.tsx:161]: that's a good point Ade, I will create a separate task for this

File: app/client/src/pages/Editor/SaaSEditor/DatasourceForm.tsx (hunk 1 above)
Comment (sneha122) [reply app/client/src/pages/Editor/SaaSEditor/DatasourceForm.tsx]: Done! please check

File: app/client/src/pages/common/datasourceAuth/index.tsx (hunk 3 above)
Comment (sneha122) [reply app/client/src/pages/common/datasourceAuth/index.tsx]: Done! please check

File: app/client/src/pages/common/datasourceAuth/index.tsx (hunk 4 above)
Comment (sneha122) [reply app/client/src/pages/common/datasourceAuth/index.tsx]: Done! please check

File: app/client/src/sagas/DatasourcesSagas.ts (hunk 5 above)
Comment (sneha122) [reply app/client/src/sagas/DatasourcesSagas.ts]: Done! please check

File: app/client/src/sagas/DatasourcesSagas.ts (hunk 6 above)
Comment (sneha122) [reply app/client/src/sagas/DatasourcesSagas.ts]: Done! please check

**Review by sumitsum — DISMISSED** [review]

**Comment by sumitsum** [comment]
Server side changes LGTM.

**Review by Irongade — DISMISSED** [review]

**Review by Irongade — APPROVED** [review]
